feat(core): schematic to migrate invalid interpolation-like markup

ayazhafiz opened the pull request:
This commit introduces a schematic that fixes interpolation-like markup
no longer accepted by the Anuglar compiler since invalidated by #39107.

Previously, the angular parser would not match markup like
`{{ 1 + 2 }` or `{{ 1 + 2 }<!-- -->}` to be partial (but malformed)
interpolations, an implementation detail that some developers may have
relied on to treat interpolation-like markup as text.

However, this kind of markup may be ambiguous, and since #39107, is not
well-formed. In such cases it is preferred to use an interpolation
directly or escape interpolation delimiters (e.g. `{{ '{{' }} 1 + 2
{{ '}}' }}`). This commit provides fixes for malformed cases so as to
ease migration for applications relying on the previous implementation.

Part of #38596

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ayazhafiz committed:
feat(core): schematic to migrate invalid interpolation-like markup

This commit introduces a schematic that fixes interpolation-like markup
no longer accepted by the Anuglar compiler since invalidated by #39107.

Previously, the angular parser would not match markup like
`{{ 1 + 2 }` or `{{ 1 + 2 }<!-- -->}` to be partial (but malformed)
interpolations, an implementation detail that some developers may have
relied on to treat interpolation-like markup as text.

However, this kind of markup may be ambiguous, and since #39107, is not
well-formed. In such cases it is preferred to use an interpolation
directly or escape interpolation delimiters (e.g. `{{ '{{' }} 1 + 2
{{ '}}' }}`). This commit provides fixes for malformed cases so as to
ease migration for applications relying on the previous implementation.

Part of #38596

ayazhafiz commented on packages/core/schematics/migrations.json:
```diff
@@ -79,6 +79,11 @@
       "description": "Updates the `initialNavigation` property for `RouterModule.forRoot`.",
       "factory": "./migrations/initial-navigation/index"
+    },
+    "migration-v12-invalid-interpolations": {
+      "version": "12",
+      "description": "Fixes interpolation-like markup not accepted by the compiler parser.",
+      "factory": "./migrations/invalid-interpolations/index"
```
Adding as a v12 migration for now, we can move this to 11.1 if you all feel like we can push this out as a bug fix

ayazhafiz committed:
fixup! feat(core): schematic to migrate invalid interpolation-like markup

JoostK commented:
Drive by comment: should we treat templates with a custom interpolation config specially somehow? Maybe skip them entirely?

ayazhafiz committed:
fixup! feat(core): schematic to migrate invalid interpolation-like markup

ayazhafiz commented:
> Drive by comment: should we treat templates with a custom interpolation config specially somehow? Maybe skip them entirely?

That's a good idea. Will do.

ayazhafiz committed:
fix(core): do not apply schematic fixes on templates with custom config

In case a user has a template that uses a custom interpolation
configuration, we should not try to apply this schematic. Trying to
predict and validate this case is very difficult to what is predictably
little value.

atscott commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+  span: AbsoluteSourceSpan;
+}
+
+/** Entry point for the V8 template variable assignment schematic. */
```
nit: update comment

ayazhafiz replied:
Done

atscott commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+  const collectedFixes: string[] = [];
+  const fixesByFile = getFixesByFile(templateVisitor.resolvedTemplates);
+
+  fixesByFile.forEach((fixes, absFilePath) => {
```
Google style: "Do not use `Array.prototype.forEach`" (use `for (... of someArr)` instead)

ayazhafiz replied:
Interesting, I thought this wouldn't work b/c tsserver complained that downLevelIteration wasn't on. I guess the build tsconfig differs somewhere...

atscott commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+
+  fixesByFile.forEach((fixes, absFilePath) => {
+    const treeFilePath = normalize(relative(basePath, absFilePath));
+    const origFileContent = tree.read(treeFilePath)?.toString();
```
nit: 
```suggestion
    const originalFileContent = tree.read(treeFilePath)?.toString();
```

ayazhafiz replied:
Done

atscott commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+      for (const {span, original} of tmplFix.invalidInterpolations) {
+        const {line, character} = getLineAndCharacterFromPosition(lineStartsMap, span.start);
+        collectedFixes.push(`${treeFilePath}@${line + 1}:${character + 1}: ${original}`);
+      }
```
Is there any reason we can't do this in the above `for (const fix of fixes)` loop? If you choose to keep these loops separate, can we align the variable names so it's easier to see that they're the same?

ayazhafiz replied:
Good idea, done :)

atscott commented on packages/core/schematics/test/invalid_interpolations_migration_spec.ts:
```diff
@@ -0,0 +1,234 @@
+
+      @Component({
+        template: \`
+          <div>{{ '{{' }} 1 + 2 {{ '}' }}</div> {{ '{{' }} 5 + 6 {{ '}' }}
```
Ah, so rather than "fixing" these by making them valid (`{{ 1 + 2 }}`), the migration maintains the current behavior by converting those to string literals. I think this is not a bad idea; if a dev wants it the other way, the log/commit log will show them exactly where these issues are.

ayazhafiz replied:
Yep, I figured this would be better so as to avoid breaking any application code.

atscott reviewed (approved):
👍 
I'm in favor of adding this migration to the 11.1 release, especially if we keep the decision to not "fix" invalid interpolations but convert them to string literals so the final template output is identical to what it is currently.

kyliau commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+
+  // Replace "{{expr}<!-- cmt -->}" with "{{ '{{' }} expr {{ '}}' }}".
+  // The former is a little-known pattern previously used to display an
+  // iterpolation literally, but that does not longer parse. The latter would
```
1. typo: interpolation
2. "but that no longer parses"

ayazhafiz replied:
Done

kyliau commented on packages/core/schematics/test/invalid_interpolations_migration_spec.ts:
```diff
@@ -0,0 +1,234 @@
+
+describe('invalid interpolation migration', () => {
+  let runner: SchematicTestRunner;
+  let host: TempScopedNodeJsSyncHost;
```
I think it's better to use `TestHost` from `@angular-devkit/core`.
`TestHost` is backed by memory, so we don't have to worry about deletions and clean ups.
@clydin Could you please confirm whether this is best practice?

clydin replied:
Instead of a `HostTree` you can use an `EmptyTree`.  The `UnitTestTree` can then be used directly to setup the test conditions. The following can be used as an example: https://github.com/angular/angular-cli/blob/e22c196c32278dee78f56088613de51bbf826131/packages/schematics/angular/migrations/update-11/add-declaration-map-compiler-option_spec.ts#L53
`TempScopedNodeJsSyncHost` was a workaround to allow testing the `tslint-fix` task (which is now deprecated).

ayazhafiz replied:
I think that the `TempScopedNodeJsSyncHost` is needed because creating a ts.Program for the schematic requires an absolute path to the tsconfig on the native filesystem:

https://github.com/angular/angular/blob/90c0772743fda18e6b4bab6feaa731038b872cbc/packages/core/schematics/utils/typescript/compiler_host.ts#L15-L37

so we need to write a temp file to disk. It looks like around 10 other schematics also use this host for testing.

clydin replied:
That's actually a larger problem that should be addressed moving forward.  The schematic tree is not necessarily in the same state as the native filesystem.  Also using `process.cwd()` is error-prone since the schematic tree will always be rooted at the workspace base path but the process execution location is arbitrary.  However, I don't think those changes should be put in place yet as I think copy-on-write support will be needed within the schematics runtime to avoid memory usage concerns with the large volume of file access this would entail.

kyliau commented on packages/core/schematics/utils/ng_component_template.ts:
```diff
@@ -85,6 +88,8 @@ export class NgComponentTemplateVisitor {
     // Walk through all component metadata properties and determine the referenced
     // HTML templates (either external or inline)
+    let interpolationConfig: InterpolationConfig = DEFAULT_INTERPOLATION_CONFIG;
+    let resolvedTemplate: Omit<ResolvedTemplate, 'interpolationConfig'>|undefined;
```
Do we need to explicitly include `undefined`?
I thought `let` variables are automatically typed as `T|undefined` under strict more.

clydin replied:
`let` allows it reassignment.  Without the `undefined` type, the compilation will error on first usage since the variable has not been assigned (assuming no assignment between declaration and use).

kyliau reviewed (approved):
Well done, this PR is **excellent**!

clydin commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+}
+
+/** Entry point for the V8 template variable assignment schematic. */
+export default function(): Rule {
```
The schematics runtime will use the default export as the entry point for a schematic.

clydin commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+  });
+
+  if (collectedFixes.length > 0) {
+    logger.info('---- Invalid Interpolation schematic ----');
```
The update process will automatically output the migration name/description.

clydin commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+  const templateVisitor = new NgComponentTemplateVisitor(typeChecker);
+  const sourceFiles = program.getSourceFiles().filter(
+      f => !f.isDeclarationFile && !program.isSourceFileFromExternalLibrary(f));
+  sourceFiles.forEach(sourceFile => templateVisitor.visitNode(sourceFile));
```
To avoid walking every TS file in the project (potentially 1000s), what do you think about narrowing the set by adding a check similar to `sourceFile.text.includes('template')`?

ayazhafiz replied:
That's a great idea, done.

JoostK commented on packages/core/schematics/migrations/invalid-interpolations/README.md:
```diff
@@ -0,0 +1,15 @@
+This schematic fixes interpolation-like markup no longer accepted by
+the Anuglar compiler since invalidated by #39107.
+
+Previously, the angular parser would not match markup like
```
```suggestion
Previously, the Angular parser would not match markup like
```

ayazhafiz replied:
Done

JoostK commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+  const invalidInterpolations: InvalidInterpolation[] = [];
+  for (const [reInterpolation, replacement] of RE_INTERPOLATIONS) {
+    // First detect invalid interpolations, then fix all of them.
+    let match;
```
```suggestion
    let match: RegExpExecArray;
```

ayazhafiz replied:
Done

JoostK commented on packages/core/schematics/utils/ng_component_template.ts:
```diff
@@ -121,15 +125,28 @@ export class NgComponentTemplateVisitor {
-        });
+        };
+      } else if (
+          propertyName === 'interpolation' && ts.isArrayLiteralExpression(property.initializer) &&
+          property.initializer.elements.length === 2 &&
+          property.initializer.elements.every(ts.isStringLiteralLike)) {
+        const [start, end] = property.initializer.elements;
```
This requires the config to be specified inline; if it's not we'll still assume it's the default configuration. I think that's an undesirable assumption to make, because I can see this being a reference to a constant that is reused across all templates.

How about we only track whether `interpolation` is specified at all, without inspecting its value? That's what's done currently anyway, since the migrations only compares to `DEFAULT_INTERPOLATION_CONFIG` reference anyway (not whether the individual start and end delimiters are `{{` and `}}`).

ayazhafiz replied:
Good catch. However, I don't think we should lose information if we don't have to (in the contrived case someone specifies `interpolation: ['{{', '}}']`). I've changed it to return `DEFAULT_INTERPOLATION_CONFIG` if the config is not specified, the custom config if the property is an array literal, or `'unknown'` otherwise.

JoostK commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+  span: AbsoluteSourceSpan;
+}
+
+/** Entry point for the V8 template variable assignment schematic. */
```
This comment is inaccurate 🤡

ayazhafiz replied:
Removed

ayazhafiz commented on packages/core/schematics/migrations/invalid-interpolations/README.md:
```diff
@@ -0,0 +1,15 @@
+## Invalid interpolation-like markup
+
+This schematic fixes interpolation-like markup no longer accepted by
+the Anuglar compiler since invalidated by #39107.
```
Done

ayazhafiz commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+import {logging, normalize} from '@angular-devkit/core';
+import {Rule, SchematicContext, SchematicsException, Tree} from '@angular-devkit/schematics';
+import {AbsoluteSourceSpan, DEFAULT_INTERPOLATION_CONFIG} from '@angular/compiler';
+import {relative} from 'path';
```
Good catch, thank you. tsserver's auto import code action didn't do what I wanted ):

ayazhafiz committed:
fixup! fix(core): do not apply schematic fixes on templates with custom config

JoostK commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,185 @@
+    // about the detected malformed interpolation-like content.
+    const lineStartsMap = computeLineStartsMap(originalFileContent);
+    const updater = tree.beginUpdate(treeFilePath);
+    for (const fix of fixes) {
```
I think that having multiple fixes here is a bit of a problem for external templates. That would happen when the same template files is used for multiple component classes, but in that case the migration only needs to be applied once (it is independent of the component class). The below `updater.remove`/`updater.insertLeft` would get out of sync for overlapping (read: identical) templates.

We should of course still support multiple fixes per file for inline templates. I don't know to what extend this is a pre-existing issue that also applies to existing migrations. We should have tests for these scenarios (multiple templates per file, and an external template that is used from multiple components)

ayazhafiz replied:
This is a great comment. I have added logic and a test to cover this case.

ayazhafiz committed:
fixup! feat(core): schematic to migrate invalid interpolation-like markup

gkalpak commented on packages/core/schematics/migrations/invalid-interpolations/BUILD.bazel:
```diff
@@ -0,0 +1,20 @@
+        "@npm//@angular-devkit/core",
+        "@npm//@angular-devkit/schematics",
+        "@npm//@types/node",
+        "@npm//typescript",
```
Is this needed here?

ayazhafiz replied:
Nope! Thanks.

gkalpak commented on packages/core/schematics/migrations.json:
```diff
@@ -79,6 +79,11 @@
       "factory": "./migrations/initial-navigation/index"
+    },
+    "migration-v12-invalid-interpolations": {
+      "version": "12",
```
Shouldn't this be `12.0.0-beta` or something?

ayazhafiz replied:
Yes, moving to 11.1.0-beta as we plan to land this in minor.

gkalpak reviewed (approved):
A couple of minor comments. Otherwise lgtm for `dev-infra`.

Reviewed-for: dev-infra

ayazhafiz committed:
fixup! feat(core): schematic to migrate invalid interpolation-like markup

ayazhafiz commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,191 @@
+  // interpolation literally, but that no longer parses. The latter would
+  // still parse.
+  [
+    /{{((?:'[^']*?'|"[^"]*?"|[^}'"])*?)}<!--.*-->}/g,
+    //                                 ^^^^^^^^^^^ match a comment b/w braces
+    `{{ '{{' }}$1{{ '}}' }}`
```
```suggestion
    /{{((?:'[^']*?'|"[^"]*?"|[^}'"])*?)}<!--(.*?)-->}/g,
    //                                 ^^^^^^^^^^^ match a comment b/w braces
    `{{ '{{' }}$1{{ '}}' }}<!-- $2 -->`
```

alxhub commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,191 @@
+          // Only include TS files that look like they might point to a template
+          // via a "template:" or "templateUrl:" key
+          f.text.includes('template'));
+  sourceFiles.forEach(sourceFile => templateVisitor.visitNode(sourceFile));
```
Prefer `for..of`

alxhub commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,191 @@
+  const collectedFixes: string[] = [];
+  const fixesByFile = getFixesByFile(templateVisitor.resolvedTemplates);
+
+  for (const [absFilePath, fixes] of fixesByFile.entries()) {
```
I don't think you need `.entries()` here - `for (const [absFilePath, fixes] of fixesByFile)` should by default iterate the entries.

alxhub commented on packages/core/schematics/migrations/invalid-interpolations/index.ts:
```diff
@@ -0,0 +1,191 @@
+}
+
+/**
+ * Returns fixes for nodes in a template containing invalid interpolations,
```
```suggestion
 * Returns fixes for nodes in templates which may contain invalid interpolations,
```

alxhub commented on packages/core/schematics/test/invalid_interpolations_migration_spec.ts:
```diff
@@ -0,0 +1,296 @@
+      @Component({
+        template: \`
+          {{ 1 + 2 }}
+          <span>{{ 3 }} or {{ '{' + "a" + '}' }}</span> or {{ "{{" + "b" + "}}" }}
```
{{ '{{ a }' }}

ayazhafiz committed:
fixup! feat(core): schematic to migrate invalid interpolation-like markup

ayazhafiz committed:
fixup! feat(core): schematic to migrate invalid interpolation-like markup

kyliau commented:
presubmit: 	http://test/OCL:343322341:BASE:343322365:1605811335397:5511c4d7

ayazhafiz commented:
Asking Andrew to confirm that the cases satisfied by this PR cover the majority of cases broken by a g3sync of https://github.com/angular/angular/pull/39107; blocking until then.

atscott commented:
Updating from TGP results: The only new failure that we didn't account for in this schematic was an open interpolation with no attempt to close it. For example:
```
<div class="additional-text">
    Use two curly braces like so
    <span class="highlight">{{</span>variable<span class="highlight">}}</span>...
  </div>
```
At the very least, let's update the breaking change not in #39107 to include directions for "printing" the brackets via `{{ '{{' }}` (and potentially remove the note about using a comment to break it up, ie `{<!-->{`)

ayazhafiz commented:
Will discuss more in our next sync (11/24 ?) and circle back.

atscott commented:
Closing since #39107 will also not be merged. The Ivy LS implementation has been out for a while and we haven't encountered issues around this edge case.
